sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_memory.asciidoc

```diff
@@ -0,0 +1,507 @@
+only a pointer to the memory without any size information.
+The functionality of `work_group_memory` is, of course, less than
+`local_accessor`, but many applications do not need the extra features.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_WORK_GROUP_MEMORY` to one of the values defined in the table
+below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== New `work_group_memory` class
+
+This extension adds the following new class:
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template<typename DataT, typename PropertyListT = empty_properties_t>
+class work_group_memory {
+ public:
+  using value_type = std::remove_all_extents_t<DataT>;
+
+  work_group_memory();
+  work_group_memory(const work_group_memory& rhs);
+  work_group_memory(handler& cgh);
+  work_group_memory(size_t num, handler& cgh);
+  work_group_memory& operator=(const work_group_memory& rhs);
+
+  operator DataT&() const;
+  const work_group_memory& operator=(const DataT& value) const;
+  DataT* operator&() const;
+
+  template<access::decorated IsDecorated = access::decorated::no>
+  multi_ptr<value_type, access::address_space::local_space, IsDecorated> get_multi_ptr() const;
+};
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+The `work_group_memory` class allocates device local memory and provides access
+to this memory from within a SYCL kernel function.
+The local memory that is allocated is shared between all work-items of a
+work-group.
+If multiple work-groups execute simultaneously, each of those work-group
+receives its own independent copy of the allocated local memory.
+
+The `work_group_memory` class may only be used in an nd-range kernel.
+Attempting to use `work_group_memory` in a single-task kernel or a simple range
+kernel produces undefined behavior.
+
+The `work_group_memory` type is a legal kernel parameter type as defined in
+section 4.12.4 "Rules for parameter passing to kernels" of the core SYCL
+specification.
+Applications typically construct an object of type `work_group_memory` in
+command group scope, pass the object as a kernel parameter, and then reference
+the object inside the kernel in order to access the device local memory that it
+contains.
+
+The `DataT` template parameter identifies the type of the objects created in
+device local memory, and this type must be one of the types that is supported
+in device code.
```

Comment:
It cannot have any constructors etc, it must be POD type, but I guess it is exactly what "this type must be one of the types that is supported in device code" means

Reply:
Actually, the type could have constructors and it doesn't need to be POD.  The limitations on `DataT` are the same as with `local_accessor`.  The next paragraph explains the effect when the type has constructors:

> If `DataT` is an implicit-lifetime type as defined in the C++ core language, `work_group_memory` implicitly creates objects of that type with indeterminate values. For other types, `work_group_memory` merely allocates uninitialized memory, and the application is responsible for constructing objects in that memory (e.g. by calling placement-new).

If `DataT` is a class with a non-trivial constructor, it is not an [implicit-lifetime](https://en.cppreference.com/w/cpp/named_req/ImplicitLifetimeType) class.  Therefore, `work_group_memory` only allocates uninitialized memory, and the program would need to call placement-new to create objects in that memory.  This is, admittedly, a weird case which few people would use, but it is technically supported.

Reply:
ok, makes sense. It gets very tricky, but the usage is not limited